gitriddance.1
.\" -*- nroff -*-
.TH gitriddance 1 "Commands"
.SH NAME
gitriddance \- closes all open issues and pull requests
.SH SYNOPSIS
.B gitriddance [comment]
.SH DESCRIPTION
.I gitriddance
closes all open issues and pull requests on GitHub. This is useful for
projects that have their own issue trackers, patch submission systems etc,
rather than relying on GitHub's, which many of us find to be clumsy,
slow, proprietary, and encouraging of drive-by pull requests of poor quality.
.PP
It should be run in a git repository that was cloned from GitHub. It
looks at the origin remote to find the repository on GitHub. All open
issues and pull requests will have a comment posted to them, and be closed.
.PP
The text of the comment is either passed as a command-line parameter,
or can be configured by setting core.gitriddance. For example:
.IP
git config core.gitriddance "Please submit patches to http://ikiwiki.info/todo/"
.PP
In order for gitriddance to log into GitHub, you need to set 
the GITHUB_USER and GITHUB_PASSWORD environment variables.
Or, set GITHUB_OAUTH_TOKEN to an oath or personal access token.
.SH AUTHOR 
Joey Hess <id@joeyh.name>


CHANGELOG
github-backup (1.20160922) unstable; urgency=medium

  * Increase base bounds to 4.8 to get mapM_ etc that works on Vector.
  * Recommended build method is now to use stack, for more reliable builds.
  * Explicitly list modules and other files in github-backup.cabal.
  * The hackage .tar.gz now omits the debian directory and may omit other
    source files in the future; the purpose of that tarball is to make
    stack/cabal install work, not to be a complete source distribution.
    Use git repository to get complete source.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Mon, 22 Aug 2016 23:57:16 -0400

github-backup (1.20160522) unstable; urgency=medium

  * Work around git weirdness in handling of relative path to GIT_INDEX_FILE
    when in a subdirectory of the repository.
  * Various updates to internal git and utility libraries shared
    with git-annex.
  * debian: Add lintian overrides for rpath.
  * debian: Bump standards-version.
  * Makefile: Pass LDFLAGS, CFLAGS, and CPPFLAGS through ghc and on to 
    ld, cc, and cpp. This lets the Debian package build with various
    hardening options. Although their benefit to a largely haskell program
    is unknown.

 -- Joey Hess <id@joeyh.name>  Sun, 22 May 2016 15:24:04 -0400

github-backup (1.20160511) unstable; urgency=medium

  * Fix build with directory-1.2.6.2.
  * github-backup.cabal: Add Setup-Depends.

 -- Joey Hess <id@joeyh.name>  Wed, 11 May 2016 13:12:12 -0400

github-backup (1.20160319) unstable; urgency=medium

  * Update to github-0.14.1.
    Building with old versions is not supported due to the massive API
    changes in this version of the github library.
  * Added support for oath tokens.

 -- Joey Hess <id@joeyh.name>  Sat, 19 Mar 2016 14:11:55 -0400

github-backup (1.20160207) unstable; urgency=medium

  * Switch to using https instead of git:// for a secure transport.
    Note that existing git:// remotes will continue to be used.
    Thanks, Robin Schneider.
  * Add upper bound on github dependency, as version 0.14 has large API
    changes and can't be used yet.
  * Various updates to internal git and utility libraries shared
    with git-annex.
  * Silence build warnings with ghc 7.10.

 -- Joey Hess <id@joeyh.name>  Sun, 07 Feb 2016 23:29:59 -0400

github-backup (1.20150807) unstable; urgency=medium

  * Added bash completion.
  * Add --no-forks flag that turns off backing up forks.
    Thanks, Phil Ruffwind.
  * Avoid nonzero exit due to temporary failures that can be retried next
    time github-backup runs, so that it can be used in a cron job with eg,
    chronic.

 -- Joey Hess <id@joeyh.name>  Fri, 07 Aug 2015 12:50:02 -0400

github-backup (1.20150618) unstable; urgency=medium

  * Fix broken configure script.

 -- Joey Hess <id@joeyh.name>  Thu, 18 Jun 2015 16:56:41 -0400

github-backup (1.20150617) unstable; urgency=medium

  * Add missing build deps for windows to cabal file.
    Thanks, Jeff Segal.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Wed, 17 Jun 2015 14:44:14 -0400

github-backup (1.20150106) unstable; urgency=medium

  * Fix build with process 1.2.1.0.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Tue, 06 Jan 2015 19:08:08 -0400

github-backup (1.20141222) unstable; urgency=medium

  * Added gitriddance(1), a utility to close all issues and pull requests,
    for repos that don't want to be bothered with GitHub's proprietary
    issue tracker.
  * gitriddance depends on github 0.13.1, which has bug fixes
    for posting comments.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Mon, 22 Dec 2014 15:30:08 -0400

github-backup (1.20141204.1) unstable; urgency=medium

  * Set myself as maintainer.

 -- James McCoy <jamessan@debian.org>  Fri, 19 Jun 2015 21:32:18 -0400

github-backup (1.20141204) unstable; urgency=high

  * Fix broken argument parser for the username|organization parameter.
    Closes: #772043

 -- Joey Hess <id@joeyh.name>  Thu, 04 Dec 2014 12:29:06 -0400

github-backup (1.20141110) unstable; urgency=medium

  * Orphaned the Debian package.
    (I continue to maintain github-backup upstream.)

 -- Joey Hess <joeyh@debian.org>  Mon, 10 Nov 2014 12:19:59 -0400

github-backup (1.20141031) unstable; urgency=medium

  * Adjust cabal file for network-uri split.
  * Avoid using optparse-applicate's argument combinator, so it will build
    with 0.11 and older too.
  * Various updates to internal git and utility libraries shared with
    git-annex.

 -- Joey Hess <joeyh@debian.org>  Fri, 31 Oct 2014 11:17:49 -0400

github-backup (1.20140831) unstable; urgency=medium

  * Fix build with github 0.10.

 -- Joey Hess <joeyh@debian.org>  Sun, 31 Aug 2014 15:31:26 -0700

github-backup (1.20140807) unstable; urgency=medium

  * Fix build with github 0.9.
  * Fix creation of github branch.

 -- Joey Hess <joeyh@debian.org>  Thu, 07 Aug 2014 22:17:22 -0400

github-backup (1.20140721) unstable; urgency=medium

  * Fix typo in fix for url parsing. Closes: #755261

 -- Joey Hess <joeyh@debian.org>  Mon, 21 Jul 2014 13:10:49 -0400

github-backup (1.20140720) unstable; urgency=medium

  * Deal with trailing slashes on github repo urls. Closes: #755261
  * Fix bug introduced by change to embedded git libraries
    in last release. Closes: #755262

 -- Joey Hess <joeyh@debian.org>  Sun, 20 Jul 2014 17:20:26 -0400

github-backup (1.20140707) unstable; urgency=medium

  * Add --exclude to skip backing up a specific repository
    when backing up a user or organization's repositories.
    Closes: #754072
  * Converted to using optparse-applicative.
  * Multiple usernames can now be specified to back up at once.

 -- Joey Hess <joeyh@debian.org>  Mon, 07 Jul 2014 23:01:03 -0400


gitriddance.hs
{- gitriddance - close all open issues and pull requests
 -
 - Copyright 2014 Joey Hess <joey@kitenet.net>
 -
 - Licensed under the GNU GPL version 3 or higher.
 -}

module Main where

import qualified GitHub.Endpoints.Repos as Github
import qualified GitHub.Endpoints.Issues as Github
import qualified GitHub.Endpoints.Issues.Comments as Github
import qualified GitHub.Data.Id as Github
import System.Environment
import Data.String
import qualified Data.Text as T

import Common
import qualified Git
import qualified Git.Construct
import qualified Git.Config
import Github.GetAuth
import Github.EnumRepos

main :: IO ()
main = do
	auth <- fromMaybe (error "Must set GITHUB_USER and GITHUB_PASSWORD")
		<$> getAuth
	r <- maybe (error "not in a git repository") Git.Config.read
		=<< Git.Construct.fromCwd
	msg <- maybe (getMsg r) id <$> (headMaybe <$> getArgs)
	case gitHubRemotes (onlyOriginRemote r) of
		[] -> error "origin does not seem to be a github repository"
		[origin] -> closeall auth origin msg
		_ -> error "somehow found multiple origin repos; this should be impossible!"

getMsg :: Git.Repo -> String
getMsg r = fromMaybe (error "core.gitriddance needs to be set to a message to use when closing issues/pull requests (or pass the message on the command line)")
	(Git.Config.getMaybe "core.gitriddance" r)

{- Limit to only having the origin remote; we don't want to affect any
 - other remotes that might be on github. -}
onlyOriginRemote :: Git.Repo -> Git.Repo
onlyOriginRemote r = r { Git.remotes = filter isorigin (Git.remotes r) }
  where
	isorigin rmt = Git.remoteName rmt == Just "origin"

closeall :: Github.Auth -> GithubUserRepo -> String -> IO ()
closeall auth (GithubUserRepo user repo) msg =
	either (oops "getting issue list") (mapM_ close)
		=<< Github.issuesForRepo' (Just auth) (fromString user) (fromString repo) [Github.Open]
  where
	oops action err = error $ "failed " ++ action ++ ": " ++ show err
	close issue = do
		let i = Github.issueNumber issue
		putStrLn $ "closing issue: " ++ T.unpack (Github.issueTitle issue)
		either (oops "posting comment") (const $ return ())
			=<< Github.createComment auth (fromString user) (fromString repo) (Github.mkId (Github.Id 0) i) (T.pack msg)
		either (oops "closing issue/pull") (const $ return ())
			=<< Github.editIssue auth (fromString user) (fromString repo) (Github.mkId (Github.Id 0) i)
				(Github.editOfIssue { Github.editIssueState = Just (T.pack "closed") } )
